### DISCUSSION TITLE
[feat] Let miners decide on block size

### DISCUSSION THREAD

**Discussion by utxo-detective** [dogecoin/dogecoin#2925] [General]
# Feature Request

**Describe the Feature Request**

There is increased demand to scale the dogecoin network to be able to support a higher volume of transactions. A simple way to do this is to increase the block size to allow for more transactions in each block. I propose we remove block size checks from consensus and let miners decide on what blocks they are willing to mine. A natural economic equilibrium will take place between miners - if a miner tries to mine too large of a block, they could risk getting orphaned by the rest of the network. If miners mine too small of blocks, they could be leaving money on the table by missed transaction fees. Note you can still validate the blockchain using SPV proofs on cheap devices (raspberry pies), but the main block producers would need to scale up in a data center if they want to compete.

Vlad from robinhood, and elon musk are in favor of scaling dogecoin thru a block size increase

https://twitter.com/vladtenev/status/1514684185243369486
https://twitter.com/elonmusk/status/1514725812544712708

**Describe Preferred Solution**

Remove `MAX_BLOCK_SERIALIZED_SIZE`, `MAX_BLOCK_WEIGHT`, `MAX_BLOCK_BASE_SIZE` from consensus rules, instead move them into policy with sane defaults for 2022 internet speed/storage capabilities. Miners can then set their own policies and a transaction fee market can emerge.

**Implementation**

My CTO @deanmlittle and I are experienced in scaling blockchains and are capable of making these changes to the source. Can make a PR if there is sufficient interest in these changes

**Comment by patricklodder** [comment]
Even though I'm not against creating a consensus ***mechanism*** rather than a hard coded limit for block size, before we even think about that, could you please elaborate why you think that the following is an improvement?

> Miners can then set their own policies and a transaction fee market can emerge.

In particular: why does Dogecoin need a fee market? Some things I observe the Dogecoin community prides itself on is:

1. Low fee capability for tipping
2. The ability to run their own nodes

Switching from a subsidy economy to a fee economy can theoretically be a target, but you also just asked for sub-satoshi fees the other day, without answering my question if you would also need fast confirmations. If you need both then I wonder why you would want a fee market: you'll risk being squeezed for transaction fees, which happens on for example Ethereum and I assume that this a likely reason why you aren't building on there.

I will respond in-depth to some other items you bring up at a later time, but I thought that this is important to discuss first, because a lot hinges on your assertion that a fee market would help. I think that this needs some discussion.

Edit: thinking of it, this is better done in the discussions section, so I am converting this to a discussion

**Reply by utxo-detective** [reply]
If there is no transaction fee market, why are there any transaction fees in the first place? If the incentive for miners is only the subsidy, can we instead make transactions w/o paying a fee to miners at all?

**Reply by adamski-t** [reply]
Please could you write a pull request for this.

**Reply by utxo-detective** [reply]
My thinking is that miners could be incentivized by a high volume of low fee transactions, but we'd need larger block sizes for this to happen.

As for the ability for users to run their own nodes, larger blocks makes that harder, but still is possible with modern hardware and w/ reasonable costs. We can also build lightweight spv nodes that follow the chain of proof of work and are able to validate transactions that matter to that specific user, as outlined in the bitcoin whitepaper

**Reply by utxo-detective** [reply]
> Please could you write a pull request for this.

I can, yes, but i wont if its not something that will get merged. Would like to see what the core devs think before just submitting code

**Reply by patricklodder** [reply]
> why are there any transaction fees in the first place?

To prevent people storing social network data on-chain 😂 - sorry, couldn't help myself ❤️

Seriously though: to disincentivize spam. This has been the way this chain works since early 2014 when it became evident that making all transactions free invites for people to abuse the chain, so the fee is there to keep it cheap to use it for-purpose (transferring magic internet moneys) but too expensive to put a lot of non-financial data on-chain. Dogecoin is a crypto currency, not a distributed database of tweets, and the fee helps keeping it that way.

> Can we instead make transactions w/o paying a fee to miners at all?

You can try convincing all the pools to do that right now, but I would argue against it and try to convince them not to, per the above.

> My thinking is that miners could be incentivized by a high volume of low fee transactions

I understand, but we'd be creating another Solana or EOS. That's not the identity of Dogecoin per-se, and there is no organic need (yet) to move in that direction: in the first 5 months since the last checkpoint, there have been 4,285,138 on-chain transactions, while we had capacity for 684,070,313 standard transactions (within the 750kb soft cap). So we're utilizing less than 1% of the available block space organically at this time.

However, during the same time, we've subsidized miners for 2,043,090,000 DOGE. By my quick estimation, at current low fee levels, we'd need to grow capacity 1400x for miners to just earn 2x what they earn today (that means we'd have 1.4GB blockspace per minute, 2TB per day) and demand (=work for nodes) would need to grow a whopping 210,000x. Alternatively we can grow it less and raise fees symmetrically somewhere between the current 0.01 and 20 DOGE per kilobyte. No matter what, it always gets more expensive, and that means you're indirectly requesting investment from either current Dogecoin users, or miners. 

This is why I am a bit skeptical of your request and why I think the use-case you [described](https://github.com/dogecoin/dogecoin/issues/2911#issuecomment-1095073329) is not a good fit for Dogecoin's chain (or any blockchain to be honest.) If all you want is integrity or proof of authenticity, then I think you're better off signing messages and creating an ultra-sharded p2p gossip network (if you really want/need p2p) than trying to completely obliterate Dogecoin's use-case as a crypto currency just to run a social network off it's infra. The changes you propose are so far-going that it's easier to not sacrifice a crypto, especially since you're looking to directly compete with the use case the celebrities you mention were proposing it for: financial transactions.


(edit: typos)

**Reply by mwilcoxnz** [reply]
The proof of work prevents spam, not the transaction fee. Transaction fees can be recycled infinitely by dishonest miners or they can simply mine empty blocks. The only way to counteract this is by removing the block size limit and for honest miners to bare the risk of orphans

**Reply by rzwicker17** [reply]
Ln b  Bo

**Reply by ReverseControl** [reply]
@mwilcoxnz I am a noob. I would like to better follow your position. Could you please explain how PoW prevents spam? It is my understanding that to prevent spam there must be a cost to spammers such that it becomes prohibitively expensive to spam. They can still do it, but it will bankrupt them. If the transactions are free, how does PoW achieve this?

Miners can and do mine empty blocks, but by far almost none of them do. Very low frequency event. This is because of the fees, not including transactions mean losing money from the fees.

The only way? Also, replace incentive mechanism in the form of fees by trust? I am confuzzled by the contradiction in terms here. It is not a question of what counteracts this behavior. It is a question of to what degree, under what assumptions, at what cost and who is paying the cost.

Looking forward to learn from you about my first question.

Thank you.

~Escanor

**Comment by victorsk2019** [comment]
@hbeckeri Not sure if this approach is still relevant today, but may I suggest looking into colored coin implementation in order to be able to insert custom metadata (e.g. social network info) into transaction using OP_RETURN opcode?  This will likely mean that your node won't accept original coin and only will work modified node versions, but this is exactly what is mentioned in _Mastering Bitcoin (p. 280)_ for projects such as yours.

**Comment by Drael64** [comment]
My completely uneducated 2 cents:

I think with blocks VERY occasionally being filled, increasing block size a little bit, might be a good discussion to have for the near future. If activity does continue to grow, or at least experience large bursts. Personally I don't see the benefit of a variable size, and a fee market, in a currency. A currency wants to default to the lowest fees possible for most transactions.

But this needs to be balanced. Dogecoin philosophically is a little more like bitcoin or litecoin than BSV or solana. The philosophy of network participation and decentralization has empowered the network, where we now have a decent number of people running nodes. And that high amount of nodes helps other potential l1 scaling solutions like maybe halving block time (which arguably would be easier to do and have more immediate impact, assuming it doesn't cause excessive orphan blocks).

IMO, if a block size increase were considered, you'd possibly want partial nodes fully implemented, and maybe also something like mimblewimbles transaction history pruning - so that it doesn't come at the expense of participation. Like basically work to do it in as decentralized a manner possible, now that at least some new approaches to this problem exist.  I repeat though, I'm a layperson and I may have things wrong!

**Comment by daank-c** [comment]
a policy driven upgradable block-size is effectively adding capitalist-politics to doge. 

Corruption and manipulation will follow suit.

**Reply by victorsk2019** [reply]
Possibly 😜 though as a developer, I'm following KIS principle whenever possible so I think having fixed block size would be a better option to reduce any undue complexities.  Also, Elon as I understand it, wants to have a larger block size for DOGE, which makes sense because the more transactions you can squeeze into a block, the faster it will be for transactions to get processed for users.  BSV, as someone already mentioned, also worked on increasing block size.  The disadvantage, as I see it, is that with larger block sizes full node data storage will be > 50GB, but pruning block storage, as it mentioned here: https://github.com/dogecoin/dogecoin/pull/2923 may help with larger block sizes?  That's my understanding of it.
